### PR TITLE
UIMasonryList

### PR DESCRIPTION
**Pull request by savelichalex**

### CODE REVIEW

File: packages/navigation/src/index.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import './useAutoFocus';
 
 export * from './UIGridList';
+// TODO: move it
```
Comment (sertony) [inline packages/navigation/src/index.ts]: Will you?)

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+function measureList<Item>(cellWidth: number, data: MasonryItem<Item>[], numOfColumns: number) {
+    const columnsHeights = new Array(numOfColumns).fill(0);
+
+    let maxHeight = 0;
+
+    const layout: Layout = {};
+
+    // Calculating offsets for columns in advance,
+    // since for each row they are simply going to
+    // be repeated agan
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]:
```suggestion
    // be repeated again
```

File: packages/navigation/src/index.ts (hunk 1 above)
Comment (savelichalex) [reply packages/navigation/src/index.ts]: I actually don't know, what do you think, should I? Or it's good enough to keep it there for now? :)

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+
+        columnsHeights[column] += height + interColumnOffset;
+
+        // eslint-disable-next-line no-param-reassign
+        layout[item.key] = {
+            height,
+            x,
+            y,
+        };
+    }
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]: Can you add a comment that this layout algo might change in the future as it has some cons as having one column way longer then another (i.e. when having several "long" images in one column)?

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+            right = middle;
+        } else {
+            left = middle;
+        }
+
+        if (right - left <= 1) {
+            if (indexes[left].y >= coord) {
+                return left;
+            }
+            return right;
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]: 👍

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+
+type MasonryCellsRefs = Record<string, React.RefObject<UIMasonryListCellRef>>;
+
+function getMasonryCellsRefs<Item>(data: MasonryItem<Item>[]) {
+    return data.reduce<MasonryCellsRefs>((acc, { key }) => {
+        if (acc[key] != null) {
+            return acc;
+        }
+        acc[key] = React.createRef<UIMasonryListCellRef>();
+        return acc;
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]: Not sure if there is any benefit, just a shorter record (maybe it's simpler to read though).
```suggestion
        if (acc[key] == null) {
            acc[key] = React.createRef<UIMasonryListCellRef>();
        }
        return acc;
```

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+    }
+
+    React.useEffect(() => {
+        cellsRefs.current = getMasonryCellsRefs(data);
+    }, [data]);
+
+    return cellsRefs.current;
+}
+
+const virtualWindowSize = 1.5;
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]: Can you move it somewhere on top and add some comment for people to understand why do we need it for when doing virtualization?

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+                },
+            } = event;
+            /**
+             * This is a core of virtualization.
+             *
+             * At the point we already prepared a sorted index
+             * which we can easily and fast traverse to find
+             * cells keys that fit into our virtual window.
+             *
+             * We use modified binary search alrorithm,
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]:
```suggestion
             * We use modified binary search algorithm,
```

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+             * next measurement.
+             * So, we have
+             * - previous top item
+             * - current top item
+             * - previous bottom item
+             * - current bottom item
+             *
+             * Since our index is sorted, we can easily find out that
+             * items between previous and current items the ones that're
+             * changed, and others shouldn't be touched.
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]: 🔥

File: packages/navigation/src/UIGridList/UIMasonryList.tsx [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+                    item={item}
+                    renderItem={renderItem}
+                    width={cellWidth}
+                    layout={layout[item.key]}
+                />
+            ));
+
+        return (
+            <ScrollView
+                // eslint-disable-next-line react/jsx-props-no-spreading
```
Comment (sertony) [inline packages/navigation/src/UIGridList/UIMasonryList.tsx]: This one is not needed anymore ;)

**Review by sertony — APPROVED** [review]
Great job! The component should be great in terms of optimization!

**Review by ghost — APPROVED** [review]

File: packages/navigation/src/index.ts (hunk 1 above)
Comment (sertony) [reply packages/navigation/src/index.ts]: Maybe it's better to move both in some kind of Lists folder?

**Review by sertony — APPROVED** [review]

**Pull request merged**